### PR TITLE
feat: Maintain originated balance

### PR DESCRIPTION
**Pull request by metacertain**
The aim of this change is to introduce originated balances, that keep track of credit operations done for the purpose of requests originated by the node itself, and to limit monetary payments to the debt showed by the originated balance. 

	This balance changes in the following ways:

		- to keep track of originated credits
			+ it decreases with the value of these credit operations
		- to be limited by the current outstanding debt to our peer shown by the accounting balance, 
			+ when it is negative it increases to the value of the current accounting balance when the accounting balance increases to a negative value closer to or above zero, this increase can not bring the originated balance above 0
		- to reflect monetary payments made with the sole purpose of compensating for the originated balance,
			+ it increases by the amount payed for by swap settlements, possibly bringing the originated balance above 0.
			+ when originated balance is above 0 it can only decrease by originated credit operations, so if by chance we have "overpayed" for this originated balance, the surplus amount is still counted as originated traffic used after the payment happened
			

In accounting, the following changes implement this:
- When credit is done, if the originated argument is true, we deduct the price from the originated balance
- When settle (time based part) or debit decreases negative balance, originated balance is reduced towards zero along with the normal balance if the negative balance would become closer to zero.
- When notifyPaymentSent decreases negative balance, the value of the settlement is added to the originated balance, even increasing it's value above zero (as the payment was solely intended to move the originated balance). 

- When sending monetary payments limiting amount to current outstanding originated debt

The PR also includes the use of originated parameter to credit in protocols 

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/1870)

### CODE REVIEW

**Review by zelig — COMMENTED** [review]
> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1870)* status: all files reviewed, 1 unresolved discussion (waiting on @metacertain)

___
*[pkg/accounting/accounting.go, line 29 at r1](https://reviewable.io/reviews/ethersphere/bee/1870#-MaZwiiA56myntijaTZn:-MaZwiiB28zVRiqBd7VE:b-juatie) ([raw file](https://github.com/ethersphere/bee/blob/dafb2cb69827b71f4e063813e3b8e5d2992bcc99/pkg/accounting/accounting.go#L29)):*
> ```Go
> 	balancesPrefix           string    = "accounting_balance_"
> 	balancesSurplusPrefix    string    = "accounting_surplusbalance_"
> 	balancesOriginatedPrefix string    = "accounting_originatedbalance_"
> ```

the word originated is enigmatic. can we use a descriptive term?

**Comment by metacertain** [comment]
> the word originated is enigmatic. can we use a descriptive term?
we concluded with @ralph-pichler this is a descriptive term, as this is the balance representing originated traffic.
however if there is a better suggestion, we can refactor

**Review by zelig — COMMENTED** [review]
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1870)* status: all files reviewed, 1 unresolved discussion (waiting on @metacertain)

**Comment by acud** [comment]
@metacertain still not really clear to me what `originated` means in this sense here. there's no linked issue, no description of the term, i agree with viktor that this is just too obscure and not descriptive. can you please explain what it means?

**Comment by metacertain** [comment]
> @metacertain still not really clear to me what `originated` means in this sense here. there's no linked issue, no description of the term, i agree with viktor that this is just too obscure and not descriptive. can you please explain what it means?

updated pr description with explanation of changes

**Review by mrekucci — CHANGES_REQUESTED** [review]
> Reviewed 2 of 6 files at r1, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1870)* status: all files reviewed, 5 unresolved discussions (waiting on @metacertain)

___
*[pkg/accounting/accounting.go, line 27 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuJuDpAvfy4PdZbivO:-MbuJuDpAvfy4PdZbivP:bqxuw5n) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L27)):*
> ```Go
> var (
> 	_                        Interface = (*Accounting)(nil)
> 	balancesPrefix           string    = "accounting_balance_"
> ```

The `string` type can be omitted (also for the vars below) and the related variables grouped.
___
*[pkg/accounting/accounting.go, line 275 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuMCaL2ZIN3oYYjmkO:-MbuMCaL2ZIN3oYYjmkP:b-v1ceha) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L275)):*
> ```Go
> 	//
> 
> 	if originated {
> ```

This can be rewritten in early return which would make the code more readable:
```golang
if !originated {
    return nil
}

...
```
___
*[pkg/accounting/accounting.go, line 277 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuMXFL-HFQRhODMxc4:-MbuMXFL-HFQRhODMxc5:b5ohpm3) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L277)):*
> ```Go
> 	if originated {
> 		originBalance, err := a.OriginatedBalance(peer)
> 		if err != nil {
> ```

No need for double `if`, just `if err != nil && !errors.Is(err, ErrPeerNoBalance)...` will do.
___
*[pkg/accounting/accounting.go, line 360 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuNPe15zmKtS4IrM2n:-MbuNPe2ApC9Awac34KY:b-c79epb) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L360)):*
> ```Go
> 		err = a.decreaseOriginatedBalanceTo(peer, oldBalance)
> 		if err != nil {
> 			a.logger.Warningf("settle: failed to decrease originated balance: %w", err)
> ```

The `%w` wouldn't work in this context, switch to `%v`
___
*[pkg/accounting/accounting.go, line 468 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuOAcj1CCkrWd_Q_uk:-MbuOAcj1CCkrWd_Q_ul:b-hixjgl) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L468)):*
> ```Go
> 
> func originatedBalanceKey(peer swarm.Address) string {
> 	return fmt.Sprintf("%s%s", balancesOriginatedPrefix, peer.String())
> ```

The `String()` method call on the `peer` is unnecessary  since `%s` formatting directive will do the same.
___
*[pkg/accounting/accounting.go, line 704 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuP0Iq8tHmRYanapUt:-MbuP0Iq8tHmRYanapUu:b-psv364) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L704)):*
> ```Go
> 	err = a.decreaseOriginatedBalanceBy(peer, amount)
> 	if err != nil {
> 		a.logger.Warningf("accounting: notifypaymentsent failed to decrease originated balance: %w", err)
> ```

As above, replace the `%w` with `%v`.
___
*[pkg/accounting/accounting.go, line 904 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuOg94B_PxDPDcL5AO:-MbuOg94B_PxDPDcL5AP:b-psv364) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L904)):*
> ```Go
> 	err = a.decreaseOriginatedBalanceTo(peer, nextBalance)
> 	if err != nil {
> 		a.logger.Warningf("increase balance: failed to decrease originated balance: %w", err)
> ```

As above, replace the `%w` with `%v`.
___
*[pkg/accounting/accounting.go, line 953 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuP5JqE4PLY-Ia7LNm:-MbuP5JqE4PLY-Ia7LNn:b-olbnp7) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L953)):*
> ```Go
> func (a *Accounting) decreaseOriginatedBalanceTo(peer swarm.Address, limit *big.Int) error {
> 
> 	zero := big.NewInt(0)
> ```

This is unnecessary since you've already declared the `zero` value above. 
___
*[pkg/accounting/accounting.go, line 958 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuPhFI-vjQsDLOFPaI:-MbuPhFI-vjQsDLOFPaJ:b-kentyf) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L958)):*
> ```Go
> 
> 	originatedBalance, err := a.OriginatedBalance(peer)
> 	if err != nil {
> ```

As above, this can be simplified into just one `if` statement.
___
*[pkg/accounting/accounting.go, line 984 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuPw4f6a8ppuRvtpJo:-MbuPw4f6a8ppuRvtpJp:b-kentyf) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/accounting.go#L984)):*
> ```Go
> 
> 	originatedBalance, err := a.OriginatedBalance(peer)
> 	if err != nil {
> ```

As above, this can be simplified into just one `if` statement.
___
*[pkg/accounting/metrics.go, line 43 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuQwD7AES6tS5qU2bh:-MbuQwD7AES6tS5qU2bi:bqjwhvo) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/metrics.go#L43)):*
> ```Go
> 			Help:      "Amount of BZZ credited to peers (potential cost of the node)",
> 		}),
> 		TotalOriginatedCreditedAmount: prometheus.NewCounter(prometheus.CounterOpts{
> ```

This can follow the position defined in the struct and so be moved to the penultimate place.
___
*[pkg/accounting/metrics.go, line 61 at r3](https://reviewable.io/reviews/ethersphere/bee/1870#-MbuRMlSEZSfbOnKP8tq:-MbuRMlSEZSfbOnKP8tr:b-lybnme) ([raw file](https://github.com/ethersphere/bee/blob/2d7d2131a23430edee3a8b4cf3e8ad1aedbe2817/pkg/accounting/metrics.go#L61)):*
> ```Go
> 			Help:      "Number of occurrences of BZZ credit events towards peers",
> 		}),
> 		OriginatedCreditEventsCount: prometheus.NewCounter(prometheus.CounterOpts{
> ```

This can follow the position defined in the struct and so be moved to the last place.

**Comment by metacertain** [comment]
attended all comments as suggested, except

> The `String()` method call on the `peer` is unnecessary since `%s` formatting directive will do the same.

this lead to unexplained symptoms in the past, wouldn't try it at the moment

> > func (a *Accounting) decreaseOriginatedBalanceTo(peer swarm.Address, limit *big.Int) error {
> > 
> > 	zero := big.NewInt(0)
> > ```
> 
> This is unnecessary since you've already declared the `zero` value above.

we removed global zero as a risk factor, it could be changed by functions receiving it

**Review by mrekucci — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1870#-:-Mbusdin0JlIEwd77o-A:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @metacertain)

File: pkg/accounting/accounting.go [hunk 1]
```diff
@@ -42,7 +43,7 @@ type Interface interface {
 	// Release releases the reserved funds.
 	Release(peer swarm.Address, price uint64)
 	// Credit increases the balance the peer has with us (we "pay" the peer).
-	Credit(peer swarm.Address, price uint64) error
+	Credit(peer swarm.Address, price uint64, orignated bool) error
```
Comment (metacertain) [inline pkg/accounting/accounting.go:46]: typo

File: pkg/accounting/accounting.go [hunk 2]
```diff
@@ -267,6 +268,41 @@ func (a *Accounting) Credit(peer swarm.Address, price uint64) error {
...
+	}
+
+	// Calculate next balance by decreasing current balance with the price we credit
+	nextOriginBalance := new(big.Int).Sub(originBalance, new(big.Int).SetUint64(price))
+
+	a.logger.Tracef("crediting peer %v with price %d, new originated balance is %d", peer, price, nextOriginBalance)
+
+	zero := big.NewInt(0)
+	// only consider negative balance for limiting originated balance
+	if nextBalance.Cmp(zero) > 0 {
```
[the remark this answers is not in the document]
Comment (metacertain) [reply pkg/accounting/accounting.go:288]: if accounting balance is < 0
and originated balance is < accounting balance 
than originated balance = accounting balance

**Review by acud — APPROVED** [review]

**Pull request merged**